### ISSUE TITLE
Scope error after serializing procedure

### ISSUE THREAD

**Issue by rdococ** [jmoenig/Snap#2265]
https://snap.berkeley.edu/snapsource/snap.html#present:Username=rdococ&ProjectName=Prototype%20programming%20with%20associative%20lists

I have a script like this. It creates several list-based objects; an 'Account' object, a 'Named' object, a 'Named Account' object which delegates its fields and methods to Account or Named, and a 'na' object which is an example of a named account.

![serialization error making objects](https://user-images.githubusercontent.com/3169568/48810386-ff22b080-ed20-11e8-95b2-63312e9124d1.png)

For some reason, after running the script that defines these variables (and they are global variables), the below script works, but if I then serialize afterwards by using the "Zoom blocks" feature, or saving and opening, I get an error like "a variable of name 'Account' is not defined in this context".

![serialization error testing objects](https://user-images.githubusercontent.com/3169568/48810389-02b63780-ed21-11e8-8f6b-2883a7d29c82.png)

This seems to be occuring in the 'Named Account' object, which holds a prototype that is a procedure that tries to access the 'Account' variable.

See the project I linked at the top for the definitions of these blocks. They don't seem to be the issue, so I haven't included their definitions here, and even if they are then there shouldn't be differences in behavior before and after serialization.

**Comment by jguille2** [comment]
Oh, yes. And this is not a simple issue.

I think there are two different things.

- In one hand (we have discussed this in other issues), there are the scope of the upvars. You use (with the "var which is" block) the upvar out of the block slots, and this is running ok, because that upvar is created as a script var. And you "complex blocks" call the scripts like a callback (accessing to the script scope). And when cached script memory is reset, this access (as a callback) is lost. The problem is "the reset" (I will comment in the next point) but also, we must take care about these upvars use.

- In the other hand, maybe we have to rethink these "resets". We have many actions (changing blocks zoom, saving, changing language...) that reset the project and cause different effects. For example, another issue will be that "transient" vars are deleted after these actions.

**Comment by brianharvey** [comment]
I agree, this reset behavior is a bug.  If an accessible (e.g., global naned) procedure makes reference to a non-global variable for which it's in scope (maybe indirectly through anonymous methods), that variable can't disappear.

I have mixed feelings about the "transient" business.  We do advertise that making a variable transient will make its value disappear when the project is saved.  So the reset behavior on a save is correct with respect to transient variables.  (And we need a programmatic way to make non-global variables transient!)  But there is no reason for a user to expect that Zoom Blocks is going to lose those values.

So if I'm understanding this correctly, there are two parts to fixing it:

1.  Make two kinds of reset, one of which does and one of which doesn't lose the values of transient variables.  Only an actual save should call the former.

2.  Neither kind of reset should lose script variables, including upvars, that are still used by some procedure.  (I.e., if a garbage collection wouldn't kill it, because it'd find a pointer to it.)

**Comment by jguille2** [comment]
Hi @brianharvey ,

In my opinion, we don't need two kind of resets. Snap! uses this "reset" (changing lang, saving projects...) just to get an small number of actions (rewrite the url or rewrite the default names of some objects...) but really, these actions are not resets.

So the idea is avoid these resets and do these actions in another way.

In the example of _transient vars_. Of course, changing lang must keep their values. But I think _saving project_ must keep them too. Yes, I know that the stored project haven't keep these vars, but I think nobody expects that saving will delete current transient vars.

**Comment by brianharvey** [comment]
Oh, I see, you want an empty value to be saved, but keeping the large value in the current session.  I guess that could work, although one could argue that losing the value in the current session reminds you that the value has to be recomputed at startup.

I'm not sure whether your other point is just not wanting to use the _name_ "reset" for what happens when you save, or if you really want to remove some code.  I don't really know how serialization works!  But I think I remember Jens saying that there may not be room to keep the runnable project and the serialized version in memory at once.
